File: FortyTwo.Entity/Models/Match.cs
﻿using FortyTwo.Shared.Models;

namespace FortyTwo.Entity.Models
{
    public class Match
    {
        public Match()
        {
            Id = Guid.NewGuid();
            CreatedOn = UpdatedOn = DateTimeOffset.UtcNow;
            CurrentGame = new Game("Game 1");
            Games = new Dictionary<Teams, List<Game>>();
            Players = new List<MatchPlayer>();
        }

        public Match(Guid id) : this() { Id = id; }

        public Guid Id { get; }
        public Game CurrentGame { get; set; }
        public Dictionary<Teams, List<Game>> Games { get; set; }
        public Dictionary<Teams, int> Scores => Games?.ToDictionary(kv => kv.Key, kv => kv.Value.Sum(g => g.Value ?? 0));
        public Teams? WinningTeam { get; set; }
        public DateTimeOffset CreatedOn { get; set; }
        public DateTimeOffset UpdatedOn { get; set; }

        // TODO: CompletedOn ??

        public List<MatchPlayer> Players { get; set; }
    }


    public static class MatchExtensions
    {
        public static void SelectNextPlayer(this Match match)
        {
            if (string.IsNullOrWhiteSpace(match.CurrentGame?.CurrentPlayerId)) return;

            var nextPlayerPosition = match.Players.First(x => x.PlayerId == match.CurrentGame.CurrentPlayerId).Position.NextPosition();

            if (match.CurrentGame.Trump == Suit.Low)
            {
                var biddingPlayerPosition = match.Players.First(x => x.PlayerId == match.CurrentGame.BiddingPlayerId).Position;

                if (biddingPlayerPosition != nextPlayerPosition && (int)biddingPlayerPosition % 2 == (int)nextPlayerPosition % 2)
                {
                    nextPlayerPosition = nextPlayerPosition.NextPosition();
                }
            }

            match.CurrentGame.CurrentPlayerId = match.Players.First(x => x.Position == nextPlayerPosition).PlayerId;
        }
    }
}

File: cloudflare/packages/rules/src/match.ts
import { nextPosition, Positions } from './positions';
import { Suit } from './suit';

export interface MatchPlayerRef {
  playerId: string;
  position: Positions;
}

// Pure-function port of C# `MatchExtensions.SelectNextPlayer`
// (FortyTwo.Entity/Models/Match.cs). Returns the next player's id instead of mutating
// a Match in place.
export function selectNextPlayer(
  currentPlayerId: string,
  biddingPlayerId: string | null,
  trump: Suit | null,
  players: MatchPlayerRef[]
): string {
  // C#: `if (string.IsNullOrWhiteSpace(match.CurrentGame?.CurrentPlayerId)) return;` — a no-op.
  if (!currentPlayerId || currentPlayerId.trim() === '') return currentPlayerId;

  let nextPlayerPosition = nextPosition(players.find((x) => x.playerId === currentPlayerId)!.position);

  if (trump === Suit.Low) {
    const biddingPlayerPosition = players.find((x) => x.playerId === biddingPlayerId)!.position;

    if (biddingPlayerPosition !== nextPlayerPosition && biddingPlayerPosition % 2 === nextPlayerPosition % 2) {
      nextPlayerPosition = nextPosition(nextPlayerPosition);
    }
  }

  return players.find((x) => x.position === nextPlayerPosition)!.playerId;
}
